server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java

```diff
@@ -199,6 +199,13 @@ private Decision shouldClusterFilter(RoutingNode node, RoutingAllocation allocat
         }
         return null;
     }
+    
+    @Override
+    public Decision canAllocateAnyShardToNode(RoutingNode node, RoutingAllocation allocation) {
+        Decision decision = shouldClusterFilter(node, allocation);
+        return decision != null && decision == Decision.NO ? decision : Decision.ALWAYS;
```

Comment:
In case where `debugDecision` is enabled, `shouldClusterFilter` will return a NO decision that is *not* the `Decision.NO` instance, and this code will turn it to ALWAYS.
I think this should rather read: `return decision == null ? Decision.ALWAYS : decision`